applications/serial_lte_modem/src/nrfcloud/slm_at_nrfcloud.c

```diff
@@ -316,6 +322,46 @@ static int do_cloud_send_msg(const char *message, int len)
 	return err;
 }
 
+static int do_cloud_send_generic_msg(const char *message, int len)
+{
+	int err;
+	char *d2c_data;
+	char *d2c_msg;
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.rx_buf = data_buf,
+		.rx_buf_len = sizeof(data_buf),
+		.fragment_size = 0
+	};
+
+	/* format JSON reply */
+	d2c_data = k_malloc(len + 1);
+	if (d2c_data == NULL) {
+		LOG_WRN("Unable to allocate buffer");
+		return -ENOMEM;
+	}
+	strncpy(d2c_data, message, len);
+	*(d2c_data + len) = '\0';
+	d2c_msg = k_malloc(sizeof(CLOUD_D2C_MSG) + strlen(d2c_data) + 1);
+	if (d2c_msg == NULL) {
+		LOG_WRN("Unable to allocate buffer");
+		k_free(d2c_data);
+		return -ENOMEM;
+	}
+	sprintf(d2c_msg, CLOUD_D2C_MSG, d2c_data);
+
+	/* send by REST API */
+	err = nrf_cloud_rest_send_device_message(&rest_ctx, device_id, d2c_msg, false, NULL);
+	if (err) {
+		LOG_ERR("nrf_cloud_rest_send_device_message failed, error: %d", err);
+	}
+	k_free(d2c_data);
+	k_free(d2c_msg);
```

Comment:
```suggestion
#define CLOUD_D2C_MSG \
	"{\"appId\":\"MESSAGE\", \"msg\":\"%.*s\"}"
...
	size_t buflen =  sizeof(CLOUD_D2C_MSG) + len + 1);

	d2c_msg = k_malloc(buflen);
	if (d2c_msg == NULL) {
		LOG_WRN("Unable to allocate buffer");
		return -ENOMEM;
	}
	snprintf(d2c_msg, buflen, CLOUD_D2C_MSG, d2c_data, len);

	/* send by REST API */
	err = nrf_cloud_rest_send_device_message(&rest_ctx, device_id, d2c_msg, false, NULL);
	if (err) {
		LOG_ERR("nrf_cloud_rest_send_device_message failed, error: %d", err);
	}
	k_free(d2c_msg);
```

To me, it looks like you are doing one allocation too many. You don't need to `strncpy()` just to add the NUL terminator. Just use `%.*s` specifier in `snprintf()` to limit the message length.

Reply:
snprintf() works magically, thanks a lot for the advice